### PR TITLE
local: handle exceptions for token checking pre-commit hook

### PR DESCRIPTION
**Pull request by jhchabran**
Follow-up to https://github.com/sourcegraph/devx-support/issues/338 

This PR introduces a mechanism to tell pre-commit to simply skip a file if it legitimately contains what looks like a token but shouldn't raise an error. 

It also fixes a very dumb bug, previously the list of files returned by `git diff` included files that got deleted, which made the script fail because `grep` obviously couldn't read them.

Typically, this happens when doing a large rebase/merge where one of those legitimate files gets added to the diff. 

## Test plan

For the exception bit: 

- Changed a file containing false positives 
- Staged the changes
- Tried to commit
- Observed the failure because a token is detected 
- Added the comment to signal the exception
- Staged the changes again
- Tried to commit 
- Observed that it got ignored as expected. 

For the deleted files, I simply re-used the branch that was provided in the support ticket and observed that it now works with the patch in this PR.  

## Preview 🤩
[Preview Link](https://docs.sourcegraph.com/@jh/handle-pre-commit-exceptions)

### CODE REVIEW

File: dev/check-tokens.sh [hunk 1]
```diff
@@ -7,12 +7,20 @@
 
 set -eu
 
-files=$(git diff --name-only --staged)
+# Select the files to inspect. We don't want to list files which are deleted, as it makes no sense
+# to look for a token being committed in those.
+#
+# So we use --diff-filter, that tells git to only include in the diff files that were
```
Comment (burmudar) [inline dev/check-tokens.sh:13]: TIL

**Review by evict — APPROVED** [review]

File: dev/check-tokens.sh [hunk 2]
```diff
@@ -7,12 +7,20 @@
...
+# - "C" copied
+# - "M" modified
+# - "R" renamed
+files=$(git diff --name-only --staged --diff-filter ACMR)
 
 function check() {
   local file="$1"
 
-  if grep -qE "s(?:g[psd]|lk)_[0-9a-fA-F]{40,}" "$file"; then
+  if ! grep -qE "(?:(?:\/\/)|(?:#)|(?:--)) ?pre-commit:ignore_sourcegraph_token$" "$file" && grep -qE "s(?:g[psd]|lk)_[0-9a-fA-F]{40,}" "$file" ; then
```
Comment (burmudar) [inline dev/check-tokens.sh:23]: image

**Review by burmudar — APPROVED** [review]
nice addition of the `// pre-commit` directive 🥳

**Pull request merged**